src/lib/util/Config.ts

```diff
@@ -110,12 +113,35 @@ export class Config {
 
 		this.typeChecker = new TypeChecker();
 		this.fileHandler = new FileHandler();
+		this.versionUpgradeManager = new VersionUpgradeManager();
 		this.sfdxScannerPath = Controller.getSfdxScannerPath();
 		this.configFilePath = path.join(this.sfdxScannerPath, CONFIG_FILE);
 		this.logger.setLevel(LoggerLevel.TRACE);
 		await this.initializeConfig();
 
 		this.initialized = true;
+		// Upgrades MUST happen AFTER the config is marked as initialized, to prevent any unexpected recursion.
+		await this.upgradeConfig();
+	}
+
+	private async upgradeConfig(): Promise<void> {
+		if (this.versionUpgradeManager.upgradeRequired(this.configContent.currentVersion)) {
+			let error: Error = null;
+			try {
+				this.configContent.currentVersion = await this.versionUpgradeManager.upgradeToLatest(this.configContent, this.configContent.currentVersion);
+			} catch (e) {
+				// Keep the error so we can rethrow it.
+				error = e;
```

Comment:
In what cases can this occur? Should we skip writing the config if an error occurs?

Reply:
Currently, it shouldn't happen ever. But theoretically, as the scripts get more complicated (or just more numerous), errors are possible.
IMO, we shouldn't skip writing the config. Suppose someone upgrades from a very old version, and five scripts run. If the last one fails, then there's no reason to not persist the changes made by the other four. That would at least enable them to upgrade to a less-old version instead of forcing them to stay on the older version.